### PR TITLE
Use video_tick for source loop instead of a thread

### PR DESCRIPTION
**Pull request by BitRate27**
- Created a ndi_video_tick_context_t to store the local variables that were in ndi_source_thread. 
- Added ndi_video_tick_context_t tc to ndi_source_t.
- Set the video_tick callback in the obs_source_info to ndi_video_tick.
- Removed thread start and stop calls.
- Changed name of ndi_source_thread to ndi_video_tick.
- Removed local variables.
- Accessed old local variables indirectly through the s->tc.
- Changed break to running = false.
- Changed continue to return.
- Changed function names and debug comments from 'thread' to 'video tick'. 
- **Caution**: Had to increase number of samples when getting audio frame when using frame_sync from 1024 to 1600. This should be a computation based on number of seconds since the last video_tick and the sample rate of the audio which I couldn't find. 
- Noticed no audio when !frame_sync, but this was not working before this change either.

### CODE REVIEW

**Comment by BitRate27** [comment]
Also moved `auto config = Config::current()` from start of old thread to just before it is used for checking tally enabled.

File: src/ndi-source.cpp [hunk 1]
```diff
@@ -115,10 +115,42 @@ typedef struct ndi_source_config_t {
...
 
+typedef struct ndi_video_tick_context_t {
+	ptz_t ptz;
+	NDIlib_tally_t tally;
+
+	obs_source_audio obs_audio_frame;
+	obs_source_frame obs_video_frame;
+
+	NDIlib_recv_create_v3_t recv_desc;
+	;
```
Comment (Trouffman) [inline src/ndi-source.cpp:126]: typo here ?

**Comment by paulpv** [comment]
@BitRate27 this is a beautiful change, but I am leaving on holiday in a few hours and need some time to think about these changes.
I probably will not be able to focus on this until I get back ~9/3.

**Comment by paulpv** [comment]
@BitRate27 Can you give a summary here of any performance metric improvements of using this technique?

**Comment by paulpv** [comment]
@BitRate27 It would be awesome to be able to A/B test this live in the same build; have both the previous thread way and this new video_tick way.

FYI, OBS link to `obs_source_info.video_tick`:
https://docs.obsproject.com/reference-sources#c.obs_source_info.video_tick

**Comment by Trouffman** [comment]
I am struggling to understand that part of the video_tick purpose, any documentation that would address the overall advantages?

Does this work when it is audio only ?

**Comment by BitRate27** [comment]
The original code used a separate thread to capture the frames and send them on to the source API for further processing by OBS. This thread was a loop constantly capturing frames and passing them along. The time between capturing frames depended on how long it took to go through the loop. This time could vary depending on processor speed, how much data to process, and various settings. The theory is that if the loop takes less that 33ms (30 fps), we are doing extra work in the loop as the frame didn't change. 

To be the most efficient possible, we should only capture and send a frame every 33ms. Fortunately OBS has a callback that is called at this interval called `video_tick`. In summary, I took the loop out of the thread function and ask OBS to call the function every 33ms.

**Comment by BitRate27** [comment]
> @BitRate27 Can you give a summary here of any performance metric improvements of using this technique?

I have done some tests on a debug version on a very old Windows box and do not see a decrease in CPU utilization. I suspect this is because the video_tick execution time is up close to the fps. So more tests to come.

**Comment by Trouffman** [comment]
> The original code used a separate thread to capture the frames and send them on to the source API for further processing by OBS. This thread was a loop constantly capturing frames and passing them along. The time between capturing frames depended on how long it took to go through the loop. This time could vary depending on processor speed, how much data to process, and various settings. The theory is that if the loop takes less that 33ms (30 fps), we are doing extra work in the loop as the frame didn't change. 
> 
> To be the most efficient possible, we should only capture and send a frame every 33ms. Fortunately OBS has a callback that is called at this interval called `video_tick`. In summary, I took the loop out of the thread function and ask OBS to call the function every 33ms. 
> 

Thanks! That make so much more sense. One of my concerns was to get as much of the of /else and update of settings out of the main loop to limit performance impact. Seems this is taking the processing out of the loop, for the same purpose :D !

**Comment by BitRate27** [comment]
@paulpv @Trouffman I am not seeing any performance increase with this change. I was also hoping this would fix the sync offset changing through a broadcast, but it didn't fix it either. More serious is that the audio is not working when not in frame_sync mode. I suggest we shelve/deny this PR. Will investigate other ways to increase performance and keeping a constant sync offset.

**Comment by paulpv** [comment]
@BitRate27 Good choice. Sometimes better code doesn't equal improvement. Thanks for trying, and keep trying.

**Pull request closed without merging**